cpp/farm_ng/core/prototools/event_log_reader.cpp

```diff
@@ -209,4 +174,55 @@ auto eventLogTimeOrderedIndex(
   return ordered_index;
 }
 
+Expected<EventLogReader> EventLogReader::fromPath(
+    std::filesystem::path const& log_path) noexcept {
+  FARM_TRY(auto, reader_impl, EventLogReaderImpl::fromPath(log_path));
+  EventLogReader reader;
+  reader.impl_ = Shared(reader_impl).sharedPtr();
+  return reader;
+}
+
+Expected<EventLogPos> EventLogReader::readNextEvent(
+    std::string* payload) noexcept {
+  FARM_TRY(auto, pair, impl_->readNextEventImpl(payload));
+
+  return EventLogPos(pair.first, pair.second, this->impl_);
+}
+
+auto EventLogReader::getIndex() noexcept -> std::vector<EventLogPos> const& {
+  if (impl_->index().empty()) {
+    auto reset_ok = reset();
+    if (!reset_ok) {
+      FARM_WARN("{}", reset_ok.error());
+      return impl_->index();
+    }
+    while (true) {
+      auto maybe_event = readNextEvent();
+
+      if (!maybe_event) {
+        break;
+      }
```

Comment:
nit: I would collapse this to `while(auto maybe_event = readNextEvent())` (or `auto&`)